incremental-font-transfer/src/uri_templates.rs

```diff
@@ -0,0 +1,682 @@
+    ///
+    /// See: <https://w3c.github.io/IFT/Overview.html#uri-templates>
+    fn append_id_digit(&mut self, id_value: &str, digit: u8) {
+        self.append(
+            *id_value
+                .len()
+                .checked_sub(digit.into())
+                .and_then(|index| id_value.as_bytes().get(index))
+                .unwrap_or(&b'_'),
+        )
+    }
+
+    // Appends a string to the output.
+    fn append_str(&mut self, value: &str) {
+        self.0.push_str(value)
+    }
+
+    // Appends a single byte to the output.
+    fn append(&mut self, byte: u8) {
+        self.0.push(byte.into());
+    }
+
+    // Appends the percent encoded representation (%XX) of a byte to the output.
+    fn append_percent_encoded(&mut self, byte: u8) {
+        self.0.push_str(&format!("%{:02X}", byte));
+    }
+}
+
+/// Stores information on how the template expansion treats each possible byte
+/// value.
+#[derive(Copy, Clone)]
+struct ByteInfo {
+    literal_class: LiteralClass,
+    is_hex_digit: bool,
+    is_varchar: bool,
+    value: u8,
+}
+
+/// Classifies each byte value [0-255] into how it is handled by literal expansion.
+#[derive(Copy, Clone)]
+enum LiteralClass {
+    Invalid,               // This byte is not allowed in a URI template
+    Percent,               // The % character starts a percent encoding
+    LiteralCopied,         // This byte should be copied directly
+    LiteralPercentEncoded, // This byte should be percent encoded and then copied.
+    StartExpression,       // { starts an expression.
+    EndExpression,         // } ends an expression.
+}
+
+impl ByteInfo {
+    fn new(class: LiteralClass, value: u8) -> Self {
+        ByteInfo {
+            literal_class: class,
+            is_hex_digit: Self::is_hexdig(value),
+            is_varchar: Self::is_varchar(value),
+            value,
+        }
+    }
+
+    /// Returns true if byte is a hexdig.
+    ///
+    /// As defined here: <https://datatracker.ietf.org/doc/html/rfc6570#section-1.5>
+    fn is_hexdig(byte: u8) -> bool {
+        DIGIT.contains(&byte) || HEX_ALPHA_UPPER.contains(&byte) || HEX_ALPHA_LOWER.contains(&byte)
+    }
+
+    /// Returns true if byte is a varchar.
+    ///
+    /// As defined here: <https://datatracker.ietf.org/doc/html/rfc6570#section-2.3>
+    fn is_varchar(byte: u8) -> bool {
+        ALPHA_LOWER.contains(&byte)
+            || ALPHA_UPPER.contains(&byte)
+            || DIGIT.contains(&byte)
+            || byte == b'.'
+            || byte == b'_'
+            || byte == b'%'
+            || byte == b'}'
+    }
+}
+
+const NUM_U8S: usize = 256;
+const ALPHA_UPPER: RangeInclusive<u8> = 0x41..=0x5A;
+const ALPHA_LOWER: RangeInclusive<u8> = 0x61..=0x7A;
+const DIGIT: RangeInclusive<u8> = 0x30..=0x39;
+const CTL_AND_SPACE: RangeInclusive<u8> = 0x00..=0x20;
+const HEX_ALPHA_UPPER: RangeInclusive<u8> = 0x41..=0x46;
+const HEX_ALPHA_LOWER: RangeInclusive<u8> = 0x61..=0x66;
```

Comment:
many of these are available as methods on `u8`.

- `u8::is_ascii_lowercase`
- `u8::is_ascii_uppercase`
- `u8::is_ascii_digit`
- `u8::is_ascii_hexdigit`